File: Salvo/Unity/Assets/Salvo/Sim/World/CollisionWorld.cs
using System.Collections.Generic;

namespace Salvo.Sim
{
    public struct TraceHit
    {
        public bool Hit;
        /// <summary>0–1 along the traced segment.</summary>
        public float Fraction;
        public Vec3 Point;
        public Vec3 Normal;
        public SurfaceKind Surface;
        public int BrushIndex;

        public static TraceHit Miss(Vec3 end) => new TraceHit
        {
            Hit = false, Fraction = 1f, Point = end, Normal = Vec3.Zero,
            Surface = SurfaceKind.Concrete, BrushIndex = -1
        };
    }

    /// <summary>What a trace is allowed to collide with.</summary>
    public enum TraceMask : byte
    {
        /// <summary>Everything solid, including invisible clip brushes. Player movement.</summary>
        Solid,
        /// <summary>Things that stop a bullet. Ignores clip brushes — you can shoot through
        /// an invisible wall, which is what makes it an invisible wall and not a real one.</summary>
        Bullets,
        /// <summary>Things that block line of sight. Ignores clip and glass.</summary>
        Sight
    }

    /// <summary>
    /// Static level geometry, and the queries the simulation runs against it.
    /// </summary>
    /// <remarks>
    /// A uniform grid over axis-aligned boxes. Not a BVH: level geometry here is brush-based
    /// and roughly uniform, the maps are tens of metres across, and a grid is simpler to get
    /// right. It is built once per map and shared across every match on that map — nothing
    /// mutates after construction, which is what makes sharing safe.
    /// </remarks>
    public sealed class CollisionWorld
    {
        private const float CellSize = 6f;

        private readonly MapBrush[] _brushes;
        private readonly List<int>[] _cells;
        private readonly Vec3 _gridMin;
        private readonly int _dimX, _dimY, _dimZ;

        public Aabb Bounds { get; }
        public int BrushCount => _brushes.Length;

        public CollisionWorld(MapDefinition map)
        {
            _brushes = map.Brushes.ToArray();
            Bounds = map.Bounds;
            _gridMin = map.Bounds.Min;

            Vec3 size = map.Bounds.Size;
            _dimX = System.Math.Max(1, (int)System.Math.Ceiling(size.X / CellSize));
            _dimY = System.Math.Max(1, (int)System.Math.Ceiling(size.Y / CellSize));
            _dimZ = System.Math.Max(1, (int)System.Math.Ceiling(size.Z / CellSize));

            _cells = new List<int>[_dimX * _dimY * _dimZ];
            for (int i = 0; i < _cells.Length; i++) _cells[i] = new List<int>();

            for (int i = 0; i < _brushes.Length; i++)
            {
                CellRange(_brushes[i].Box, out int x0, out int y0, out int z0,
                                            out int x1, out int y1, out int z1);
                for (int z = z0; z <= z1; z++)
                    for (int y = y0; y <= y1; y++)
                        for (int x = x0; x <= x1; x++)
                            _cells[CellIndex(x, y, z)].Add(i);
            }
        }

        private int CellIndex(int x, int y, int z) => (z * _dimY + y) * _dimX + x;

        private void CellRange(Aabb box, out int x0, out int y0, out int z0,
                                         out int x1, out int y1, out int z1)
        {
            Vec3 lo = box.Min - _gridMin;
            Vec3 hi = box.Max - _gridMin;
            x0 = SalvoMath.Clamp((int)(lo.X / CellSize), 0, _dimX - 1);
            y0 = SalvoMath.Clamp((int)(lo.Y / CellSize), 0, _dimY - 1);
            z0 = SalvoMath.Clamp((int)(lo.Z / CellSize), 0, _dimZ - 1);
            x1 = SalvoMath.Clamp((int)(hi.X / CellSize), 0, _dimX - 1);
            y1 = SalvoMath.Clamp((int)(hi.Y / CellSize), 0, _dimY - 1);
            z1 = SalvoMath.Clamp((int)(hi.Z / CellSize), 0, _dimZ - 1);
        }

        private static bool Accepts(in MapBrush brush, TraceMask mask) => mask switch
        {
            TraceMask.Solid => true,
            TraceMask.Bullets => !brush.IsClip && brush.BlocksBullets,
            TraceMask.Sight => !brush.IsClip && brush.BlocksSight,
            _ => true
        };

        /// <summary>Brush indices whose grid cells overlap a box. Duplicates removed.</summary>
        public void CandidatesInto(Aabb box, List<int> results, HashSet<int> scratch)
        {
            results.Clear();
            scratch.Clear();
            CellRange(box, out int x0, out int y0, out int z0, out int x1, out int y1, out int z1);
            for (int z = z0; z <= z1; z++)
                for (int y = y0; y <= y1; y++)
                    for (int x = x0; x <= x1; x++)
                    {
                        List<int> cell = _cells[CellIndex(x, y, z)];
                        for (int i = 0; i < cell.Count; i++)
                            if (scratch.Add(cell[i])) results.Add(cell[i]);
                    }
        }

        [System.ThreadStatic] private static List<int> _traceCandidates;
        [System.ThreadStatic] private static HashSet<int> _traceScratch;

        private static void EnsureScratch()
        {
            _traceCandidates ??= new List<int>(64);
            _traceScratch ??= new HashSet<int>();
        }

        /// <summary>A line trace against the level.</summary>
        public TraceHit Trace(Vec3 start, Vec3 end, TraceMask mask = TraceMask.Solid)
        {
            Vec3 delta = end - start;
            float distance = delta.Length;
            if (distance < 1e-5f) return TraceHit.Miss(end);
            Vec3 direction = delta / distance;

            EnsureScratch();
            var sweep = new Aabb(start, end).Expanded(new Vec3(0.05f, 0.05f, 0.05f));
            CandidatesInto(sweep, _traceCandidates, _traceScratch);

            TraceHit best = TraceHit.Miss(end);
            best.Fraction = 1f;

            for (int i = 0; i < _traceCandidates.Count; i++)
            {
                int index = _traceCandidates[i];
                if (!Accepts(_brushes[index], mask)) continue;
                if (!_brushes[index].Box.Raycast(start, direction, distance,
                                                 out float hitDistance, out Vec3 normal)) continue;

                float fraction = hitDistance / distance;
                if (fraction < best.Fraction)
                {
                    best = new TraceHit
                    {
                        Hit = true,
                        Fraction = fraction,
                        Point = start + direction * hitDistance,
                        Normal = normal,
                        Surface = _brushes[index].Surface,
                        BrushIndex = index
                    };
                }
            }
            return best;
        }

        [System.ThreadStatic] private static List<int> _sweepCandidates;
        [System.ThreadStatic] private static HashSet<int> _sweepScratch;

        /// <summary>
        /// Sweeps an axis-aligned box from one centre to another and reports the first
        /// blocking contact.
        /// </summary>
        /// <remarks>
        /// This is a Minkowski sweep: expanding each brush by the box's half extents turns
        /// box-versus-box into ray-versus-box, so the same slab test that serves bullets
        /// serves the player capsule. It has its own scratch buffers rather than sharing
        /// <see cref="Trace"/>'s, because movement code legitimately calls both and a shared
        /// buffer would be cleared underneath the caller.
        ///
        /// If the box starts already overlapping a brush the slab test reports a zero-distance
        /// hit with no meaningful normal. That is reported as <c>startedSolid</c> rather than
        /// silently treated as a wall, because the two need opposite responses: a wall is slid
        /// along, a penetration must be pushed out of first.
        /// </remarks>
        public TraceHit SweepBox(Vec3 centre, Vec3 halfExtents, Vec3 target,
                                 out bool startedSolid, TraceMask mask = TraceMask.Solid)
        {
            startedSolid = false;
            Vec3 delta = target - centre;
            float distance = delta.Length;
            var miss = TraceHit.Miss(target);
            miss.Fraction = 1f;
            if (distance < 1e-6f) return miss;
            Vec3 direction = delta / distance;

            _sweepCandidates ??= new List<int>(64);
            _sweepScratch ??= new HashSet<int>();

            var region = new Aabb(centre - halfExtents, centre + halfExtents);
            region = new Aabb(Vec3.Min(region.Min, target - halfExtents),
                              Vec3.Max(region.Max, target + halfExtents));
            CandidatesInto(region, _sweepCandidates, _sweepScratch);

            TraceHit best = miss;
            for (int i = 0; i < _sweepCandidates.Count; i++)
            {
                int index = _sweepCandidates[i];
                if (!Accepts(_brushes[index], mask)) continue;

                Aabb expanded = _brushes[index].Box.Expanded(halfExtents);
                if (!expanded.Raycast(centre, direction, distance,
                                      out float hitDistance, out Vec3 normal)) continue;

                // A zero normal means the slab test never crossed a face on the way in. Two
                // very different situations produce that, and they need opposite responses:
                // the box genuinely started inside the brush (push out), or it started exactly
                // on a face and is leaving (ignore — it is grazing, not colliding). Asking the
                // geometry directly is the only way to tell them apart.
                if (normal == Vec3.Zero)
                {
                    if (StrictlyInside(expanded, centre)) startedSolid = true;
                    continue;
                }

                float fraction = hitDistance / distance;
                if (fraction < best.Fraction)
                {
                    best = new TraceHit
                    {
                        Hit = true,
                        Fraction = SalvoMath.Clamp01(fraction),
                        Point = centre + direction * hitDistance,
                        Normal = normal,
                        Surface = _brushes[index].Surface,
                        BrushIndex = index
                    };
                }
            }
            return best;
        }

        public bool HasLineOfSight(Vec3 from, Vec3 to) => !Trace(from, to, TraceMask.Sight).Hit;

        public bool Overlaps(Aabb box, TraceMask mask = TraceMask.Solid)
        {
            EnsureScratch();
            CandidatesInto(box, _traceCandidates, _traceScratch);
            for (int i = 0; i < _traceCandidates.Count; i++)
            {
                int index = _traceCandidates[i];
                if (!Accepts(_brushes[index], mask)) continue;
                if (_brushes[index].Box.Intersects(box)) return true;
            }
            return false;
        }

        /// <summary>
        /// True when <paramref name="point"/> is inside <paramref name="box"/> by more than a
        /// hair on every axis. Merely touching a face does not count.
        /// </summary>
        private static bool StrictlyInside(Aabb box, Vec3 point)
        {
            const float epsilon = 1e-4f;
            return point.X > box.Min.X + epsilon && point.X < box.Max.X - epsilon
                && point.Y > box.Min.Y + epsilon && point.Y < box.Max.Y - epsilon
                && point.Z > box.Min.Z + epsilon && point.Z < box.Max.Z - epsilon;
        }

        /// <summary>
        /// Pushes a box out of any geometry it is overlapping and returns the corrected centre.
        /// </summary>
        /// <remarks>
        /// The safety net under the movement model, and it earns its place. Sweeping alone is
        /// not enough in practice: floating-point error at a wall/floor seam, a stance change
        /// that grows the box into a ceiling, or a spawn placed a millimetre inside a brush all
        /// leave the player slightly embedded. Without a push-out, the next sweep starts solid,
        /// the model has nothing sensible to do, and the player falls through the level — which
        /// is precisely the class of bug that only ever gets found by a player mid-match.
        ///
        /// <para>Resolution is along the axis of least penetration, deepest overlap first, a
        /// few times over. Least-penetration is what makes a player embedded in a floor pop up
        /// rather than sideways. The iteration count is small and fixed because this must cost
        /// the same every tick: an unbounded loop here would be a stutter under exactly the
        /// conditions that cause it.</para>
        /// </remarks>
        public Vec3 Depenetrate(Vec3 centre, Vec3 halfExtents, TraceMask mask = TraceMask.Solid,
                                int iterations = 4)
        {
            _sweepCandidates ??= new List<int>(64);
            _sweepScratch ??= new HashSet<int>();

            for (int pass = 0; pass < iterations; pass++)
            {
                var box = new Aabb(centre - halfExtents, centre + halfExtents);
                CandidatesInto(box, _sweepCandidates, _sweepScratch);

                float deepest = 0f;
                Vec3 push = Vec3.Zero;

                for (int i = 0; i < _sweepCandidates.Count; i++)
                {
                    int index = _sweepCandidates[i];
                    if (!Accepts(_brushes[index], mask)) continue;
                    Aabb brush = _brushes[index].Box;
                    if (!brush.Intersects(box)) continue;

                    // Overlap on each axis; the smallest is the cheapest way out.
                    float ox = System.Math.Min(box.Max.X, brush.Max.X) - System.Math.Max(box.Min.X, brush.Min.X);
                    float oy = System.Math.Min(box.Max.Y, brush.Max.Y) - System.Math.Max(box.Min.Y, brush.Min.Y);
                    float oz = System.Math.Min(box.Max.Z, brush.Max.Z) - System.Math.Max(box.Min.Z, brush.Min.Z);
                    if (ox <= 0f || oy <= 0f || oz <= 0f) continue;

                    float overlap;
                    Vec3 direction;
                    if (oy <= ox && oy <= oz)
                    {
                        overlap = oy;
                        direction = centre.Y >= brush.Centre.Y ? Vec3.Up : -Vec3.Up;
                    }
                    else if (ox <= oz)
                    {
                        overlap = ox;
                        direction = centre.X >= brush.Centre.X ? Vec3.Right : -Vec3.Right;
                    }
                    else
                    {
                        overlap = oz;
                        direction = centre.Z >= brush.Centre.Z ? Vec3.Forward : -Vec3.Forward;
                    }

                    if (overlap > deepest)
                    {
                        deepest = overlap;
                        push = direction * overlap;
                    }
                }

                if (deepest <= 0f) break;
                // A sliver beyond the overlap, so the result is clear of the brush rather than
                // exactly touching it — exactly touching is what caused this in the first place.
                centre += push + push.Normalized * 1e-3f;
            }
            return centre;
        }

        /// <summary>Height of the floor below a point, or null if there is nothing under it
        /// within <paramref name="maxDrop"/>.</summary>
        public float? GroundHeight(Vec3 point, float maxDrop = 30f)
        {
            TraceHit hit = Trace(point, point + new Vec3(0f, -maxDrop, 0f), TraceMask.Solid);
            return hit.Hit ? hit.Point.Y : (float?)null;
        }

        public SurfaceKind SurfaceBelow(Vec3 point, float maxDrop = 3f)
        {
            TraceHit hit = Trace(point, point + new Vec3(0f, -maxDrop, 0f), TraceMask.Solid);
            return hit.Hit ? hit.Surface : SurfaceKind.Concrete;
        }
    }
}

File: CriticalStrike/Sources/CriticalStrikeCore/World/CollisionWorld.swift
import Foundation

public struct TraceResult: Sendable {
    public var hit: Bool
    public var fraction: Float       // 0...1 along the trace
    public var point: Vec3
    public var normal: Vec3
    public var surface: SurfaceKind
    public var brushIndex: Int
    public var startedSolid: Bool

    public static func miss(end: Vec3) -> TraceResult {
        TraceResult(hit: false, fraction: 1, point: end, normal: .zero,
                    surface: .concrete, brushIndex: -1, startedSolid: false)
    }
}

/// Uniform-grid broadphase over the map's brushes. Static geometry only: players and
/// projectiles are traced against separately, which keeps this structure immutable and
/// safe to share across threads (the bot threads raycast it constantly).
public final class CollisionWorld: @unchecked Sendable {
    public let brushes: [MapBrush]
    public let bounds: AABB

    private let cellSize: Float
    private let gridMin: Vec3
    private let gridDim: (x: Int, y: Int, z: Int)
    private var cells: [[Int32]]

    public init(map: MapData, cellSize: Float = 6) {
        self.brushes = map.brushes
        self.bounds = map.bounds.expanded(by: Vec3(2, 2, 2))
        self.cellSize = cellSize
        self.gridMin = bounds.min
        let size = bounds.size
        let dx = max(1, Int(ceil(size.x / cellSize)))
        let dy = max(1, Int(ceil(size.y / cellSize)))
        let dz = max(1, Int(ceil(size.z / cellSize)))
        self.gridDim = (dx, dy, dz)
        self.cells = Array(repeating: [], count: dx * dy * dz)
        buildGrid()
    }

    private func cellIndex(_ ix: Int, _ iy: Int, _ iz: Int) -> Int {
        (iz * gridDim.y + iy) * gridDim.x + ix
    }

    private func clampedCell(_ p: Vec3) -> (Int, Int, Int) {
        let local = p - gridMin
        return (MathUtil.clamp(Int(local.x / cellSize), 0, gridDim.x - 1),
                MathUtil.clamp(Int(local.y / cellSize), 0, gridDim.y - 1),
                MathUtil.clamp(Int(local.z / cellSize), 0, gridDim.z - 1))
    }

    private func buildGrid() {
        for (i, brush) in brushes.enumerated() {
            let (x0, y0, z0) = clampedCell(brush.box.min)
            let (x1, y1, z1) = clampedCell(brush.box.max)
            for z in z0...z1 {
                for y in y0...y1 {
                    for x in x0...x1 {
                        cells[cellIndex(x, y, z)].append(Int32(i))
                    }
                }
            }
        }
    }

    /// All brush indices whose cell overlaps `box`.
    public func candidates(in box: AABB) -> [Int] {
        let (x0, y0, z0) = clampedCell(box.min)
        let (x1, y1, z1) = clampedCell(box.max)
        var seen = Set<Int32>()
        var out: [Int] = []
        out.reserveCapacity(16)
        for z in z0...z1 {
            for y in y0...y1 {
                for x in x0...x1 {
                    for idx in cells[cellIndex(x, y, z)] where seen.insert(idx).inserted {
                        out.append(Int(idx))
                    }
                }
            }
        }
        return out
    }

    // MARK: - Ray traces

    /// Line trace against static geometry. `ignoreNonSolid` skips glass and other
    /// bullet-permeable brushes (used by the ballistics pass, which handles them itself).
    public func trace(from start: Vec3, to end: Vec3,
                      mask: TraceMask = .solid) -> TraceResult {
        let delta = end - start
        let distance = delta.length
        guard distance > 1e-5 else { return .miss(end: end) }
        let ray = Ray(origin: start, direction: delta / distance)

        var best = TraceResult.miss(end: end)
        best.fraction = 1

        let sweepBox = AABB(min: Vec3(min(start.x, end.x), min(start.y, end.y), min(start.z, end.z)),
                            max: Vec3(max(start.x, end.x), max(start.y, end.y), max(start.z, end.z)))
            .expanded(by: Vec3(0.05, 0.05, 0.05))

        for i in candidates(in: sweepBox) {
            let brush = brushes[i]
            guard mask.accepts(brush) else { continue }
            guard let (t, normal) = brush.box.raycast(ray, maxDistance: distance) else { continue }
            let f = t / distance
            if f < best.fraction {
                best = TraceResult(hit: true, fraction: f, point: ray.point(at: t),
                                   normal: normal, surface: brush.surface,
                                   brushIndex: i, startedSolid: false)
            }
        }
        return best
    }

    /// True when nothing blocks vision between two points (used by AI and by smoke checks).
    public func hasLineOfSight(from: Vec3, to: Vec3) -> Bool {
        !trace(from: from, to: to, mask: .vision).hit
    }

    // MARK: - Box sweeps (player movement)

    /// Swept AABB against the world, returning the first blocking contact.
    /// Uses conservative advancement: cheap, stable, and good enough at 64Hz.
    public func sweep(box: AABB, delta: Vec3, mask: TraceMask = .solid) -> TraceResult {
        let distance = delta.length
        guard distance > 1e-6 else { return .miss(end: box.center) }
        let dir = delta / distance

        let expandedArea = box.union(box.offset(by: delta)).expanded(by: Vec3(0.1, 0.1, 0.1))
        var bestFraction: Float = 1
        var bestNormal = Vec3.zero
        var bestSurface = SurfaceKind.concrete
        var bestIndex = -1
        var startedSolid = false

        let half = box.extents
        let center = box.center

        for i in candidates(in: expandedArea) {
            let brush = brushes[i]
            guard mask.accepts(brush) else { continue }
            // Minkowski: inflate the brush by our half-extents and ray-trace the center point.
            let inflated = brush.box.expanded(by: half)
            if inflated.contains(center) {
                startedSolid = true
                continue
            }
            guard let (t, normal) = inflated.raycast(Ray(origin: center, direction: dir),
                                                     maxDistance: distance) else { continue }
            let f = max(0, t / distance)
            if f < bestFraction {
                bestFraction = f
                bestNormal = normal
                bestSurface = brush.surface
                bestIndex = i
            }
        }

        if bestIndex < 0 {
            return TraceResult(hit: false, fraction: 1, point: center + delta, normal: .zero,
                               surface: .concrete, brushIndex: -1, startedSolid: startedSolid)
        }
        // Back off a hair so we never end the frame embedded in a surface.
        let safeFraction = max(0, bestFraction - 0.001)
        return TraceResult(hit: true, fraction: safeFraction, point: center + dir * (distance * safeFraction),
                           normal: bestNormal, surface: bestSurface, brushIndex: bestIndex,
                           startedSolid: startedSolid)
    }

    public func overlaps(box: AABB, mask: TraceMask = .solid) -> Bool {
        for i in candidates(in: box) where mask.accepts(brushes[i]) {
            if brushes[i].box.intersects(box) { return true }
        }
        return false
    }

    /// Drops a point onto the first surface below it. Used by spawn validation and nav baking.
    public func groundHeight(below p: Vec3, maxDrop: Float = 30) -> Float? {
        let result = trace(from: p, to: p + Vec3(0, -maxDrop, 0), mask: .solid)
        return result.hit ? result.point.y : nil
    }

    /// Surface material directly under a point (footstep audio).
    public func surface(below p: Vec3) -> SurfaceKind {
        let r = trace(from: p + Vec3(0, 0.2, 0), to: p + Vec3(0, -1.2, 0), mask: .solid)
        return r.hit ? r.surface : .concrete
    }
}

public struct TraceMask: OptionSet, Sendable {
    public let rawValue: Int
    public init(rawValue: Int) { self.rawValue = rawValue }

    public static let solid = TraceMask(rawValue: 1 << 0)      // blocks movement
    public static let bullets = TraceMask(rawValue: 1 << 1)    // blocks bullets
    public static let vision = TraceMask(rawValue: 1 << 2)     // blocks line of sight
    public static let includeClip = TraceMask(rawValue: 1 << 3)

    func accepts(_ brush: MapBrush) -> Bool {
        if brush.isClip {
            // Player-clip brushes stop movement but not bullets or sight.
            return contains(.solid)
        }
        if contains(.bullets) && !brush.blocksBullets { return false }
        if contains(.vision) && !brush.blocksVision { return false }
        return true
    }
}
